tests/eve_tests/unit_tests/test_type_definitions.py

```diff
@@ -54,12 +54,65 @@ def test_symbol_types():
 
 class TestSourceLocation:
     def test_valid_position(self):
-        eve.type_definitions.SourceLocation(line=1, column=1, source="source")
+        eve.type_definitions.SourceLocation(line=1, column=1, source="source.py")
 
     def test_invalid_position(self):
         with pytest.raises(pydantic.ValidationError):
-            eve.type_definitions.SourceLocation(line=1, column=-1, source="source")
+            eve.type_definitions.SourceLocation(line=1, column=-1, source="source.py")
 
     def test_str(self):
-        loc = eve.type_definitions.SourceLocation(line=1, column=1, source="source")
-        assert str(loc) == "<source: Line 1, Col 1>"
+        loc = eve.type_definitions.SourceLocation(line=1, column=1, source="dir/source.py")
+        assert str(loc) == "<'dir/source.py': Line 1, Col 1>"
+
+        loc = eve.type_definitions.SourceLocation(
+            line=1, column=1, source="dir/source.py", end_line=2
+        )
+        assert str(loc) == "<'dir/source.py': Line 1, Col 1 to Line 2>"
+
+        loc = eve.type_definitions.SourceLocation(
+            line=1, column=1, source="dir/source.py", end_line=2, end_column=2
+        )
+        assert str(loc) == "<'dir/source.py': Line 1, Col 1 to Line 2, Col 2>"
+
+    def test_construction_from_ast(self):
+        import ast
+
+        ast_node = ast.parse("a = b + 1").body[0]
+        loc = eve.type_definitions.SourceLocation.from_AST(ast_node, "source.py")
+        print("\n", ast_node.end_lineno)
+        print(loc)
+
+        assert loc.line == ast_node.lineno
+        assert loc.column == ast_node.col_offset + 1
+        assert loc.source == "source.py"
```

Comment:
`.end_line` and `.end_column` not yet tested

Reply:
Oops, tests have been updated to address this.